### ISSUE TITLE
FR: 24 hours format on admin console last seen

### ISSUE THREAD

**Issue by sIRwa2** [tailscale/tailscale#7563]

**Comment by DentonGentry** [comment]
Could you explain the reasoning?
Is it personal preference for 24 hour format, or is there something which breaks?

**Comment by sIRwa2** [comment]
hi yes, its 100% personal preference. so put it the the low hanging fruit section under "nice to have" :)

**Comment by mihaip** [comment]
@tailscale/frontend not sure if we can make `date-fns` do 24 hour formatting if the user is in a locale that prefers that.

**Comment by sonovawolf** [comment]
I've seen this as a setting more than based on location, but either way it seems like a quality of life improvement.

**Comment by rosszurowski** [comment]
One thing we can do here is swap our current date formatting implementation from `date-fns` to use the new [`Intl.DateTimeFormat`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat) objects. I believe they either default to, or can be set to use your computer locale's time options. So if your computer was set to show 24h time, our admin console would show it too.

Would that'd be sufficient here?

**Comment by sIRwa2** [comment]
i think that could work.

**Comment by wsw70** [comment]
I was about to add this FR. The 24 hours format is common in the world (https://en.wikipedia.org/wiki/Date_and_time_representation_by_country#/media/File:12_24_Hours_World_Map.svg) and I would say that this is more than a "nuisance" (if I read the badge correctly). 

This is especially true in technical contexts where people are more used to ISO formatted dates, timezones etc.

It would really be a neat addition!

This is BTW also true for the date format `Aug 20, 2024` - we would say or write `20 Aug 2024` (the further you are in the expression, the less precise it becomes).

**Comment by cloveras** [comment]
I'd also like 24h format (like most people not he planet?), and even managed to make my own little duplicate issue for this (it's closed already): https://github.com/tailscale/tailscale/issues/14480

**Comment by marco-m** [comment]
I think it is important to realize that the lack of a 24h time format is not a simple "nuisance".
Instead, it is about respecting the "locale" of the users: the AM/PM time format is alien to almost everybody that don't live in the USA.

@DentonGentry (or anybody in Tailscale that does triage, should I mention somebody else?) could you please reconsider the priority of this request? Thank you!

**Comment by sIRwa2** [comment]
Jul 14, 2025 at 10:11 AM GMT+2 is nice for logs, not humans it is as friendly as: whenthestreetlightsturnon+1

**Comment by rodrigotellom** [comment]
> I think it is important to realize that the lack of a 24h time format is not a simple "nuisance". Instead, it is about respecting the "locale" of the users: the AM/PM time format is alien to almost everybody that don't live in the USA.

@marco-m I agree that different countries my have different conventions, but, for what it's worth, 12-hour time (AM/PM) is pretty common in LatAm at least. Using 12-hour time or 24-hour time is almost personal preference.

I see more of an issue using MM/DD (which is mostly used in the US) over DD/MM.

**Comment by sIRwa2** [comment]
> > I think it is important to realize that the lack of a 24h time format is not a simple "nuisance". Instead, it is about respecting the "locale" of the users: the AM/PM time format is alien to almost everybody that don't live in the USA.
> 
> [@marco-m](https://github.com/marco-m) I agree that different countries my have different conventions, but, for what it's worth, 12-hour time (AM/PM) is pretty common in LatAm at least. Using 12-hour time or 24-hour time is almost personal preference.
> 
> I see more of an issue using MM/DD (which is mostly used in the US) over DD/MM.

ISO 8601

**Comment by Synt3x** [comment]
I'm also affected by this.
I need 24h clock can't read the last online state.

**Comment by alexwlchan** [comment]
Hi everyone! Thanks for the passion and patience on this issue.

We've just deployed a change to the admin console that will switch to 24-hour formatting based on your browser language – for example, I see 16:45 rather than 4:45 PM because my browser sets `en-GB`. I'm going to close this as done, but please do open a new ticket if you have other issues in the console, date-related or otherwise.

<details>
<summary>Notes on the fix</summary>

We retrieve your locale with the following function:

```typescript
export const getLocale = (): string =>
  navigator.userLanguage ||
  (navigator.languages &&
    navigator.languages.length > 0 &&
    navigator.languages[0]) ||
  navigator.language ||
  navigator.browserLanguage ||
  navigator.systemLanguage ||
  "en-US"
```

then we pass the output to another function which decides whether to use 24-hour formatting:

```typescript
function uses24HourClock(locale: string): boolean {
  const localeFormat = new Intl.DateTimeFormat(locale, { hour: "numeric" })
  return localeFormat.resolvedOptions().hourCycle === "h23"
}
```

</details>

**Comment by Peter-Petrik** [comment]
There are those of us who prefer to see the `en-US` spelling in our browsers, and also prefer to see 24-hour clock format. Can users have a profile setting instead of the solution (incorrectly) deciding the preference for them?

**Comment by wsw70** [comment]
> There are those of us who prefer to see the `en-US` spelling in our browsers, and also prefer to see 24-hour clock format.

I am in this situation. I want to have everything in English (I am French) but with *normal* hours and date formats :)

**Comment by marco-m** [comment]
Hello @alexwlchan,

thanks for the fix. Unfortunately it is not enough, as both @Peter-Petrik and @wsw70 just above me explain very well:

There are a lot of us (1) non-native English speakers, that (2) prefer `en-US` spelling and (3) still think that 24h format is the sane way to think and communicate about time.

EDIT Opened https://github.com/tailscale/tailscale/issues/19032

**Comment by wsw70** [comment]
>  still think that 24h format is the sane way to think and communicate about time.

And dates. Dates.

**Comment by alexwlchan** [comment]
Thank you for opening another ticket (I was about to reopen this one, but I'll leave it closed); I'll think about the best way to handle that separately, but I can't promise a timeframe for picking it up.

**Comment by marco-m** [comment]
Thank you @alexwlchan I appreciate it :-)
